jacodb-taint-configuration/src/main/kotlin/org/jacodb/configuration/TaintConfigurationFeature.kt

```diff
@@ -0,0 +1,449 @@
+                to.mapNotNull { snd ->
+                    if (fst == snd) return@mapNotNull null
+
+                    CopyAllMarks(fst, snd)
+                }
+            }
+        }
+
+        override fun visit(action: CopyMark): List<Action> {
+            val from = specializePosition(method, action.from)
+            val to = specializePosition(method, action.to)
+
+            return from.flatMap { fst ->
+                to.mapNotNull { snd ->
+                    if (fst == snd) return@mapNotNull null
+
+                    action.copy(from = fst, to = snd)
+                }
+            }
+        }
+
+        override fun visit(action: AssignMark): List<Action> =
+            specializePosition(method, action.position).map { action.copy(position = it) }
+
+        override fun visit(action: RemoveAllMarks): List<Action> =
+            specializePosition(method, action.position).map { action.copy(position = it) }
+
+        override fun visit(action: RemoveMark): List<Action> =
+            specializePosition(method, action.position).map { action.copy(position = it) }
+    }
+
+    private inner class ConditionSpecializer(val method: JcMethod) : ConditionVisitor<Condition> {
+        override fun visit(condition: And): Condition = mkAnd(condition.args.map { it.accept(this) })
+
+        override fun visit(condition: Or): Condition = mkOr(condition.args.map { it.accept(this) })
+
+        override fun visit(condition: Not): Condition = Not(condition.condition.accept(this))
+
+        override fun visit(condition: IsConstant): Condition =
+            mkOr(specializePosition(method, condition.position).map { condition.copy(position = it) })
+
+        override fun visit(condition: IsType): Condition {
+            val position = specializePosition(method, condition.position)
+
+            val typeMatcher = condition.typeMatcher
+            if (typeMatcher is AnyTypeMatcher) {
+                return mkOr(position.map { ConstantTrue })
+            }
+
+            if (typeMatcher is PrimitiveNameMatcher) {
+                val types = primitiveTypes(method).filter { typeMatcher.matches(it.typeName) }
+                return mkOr(types.flatMap { type -> position.map { TypeMatches(it, type) } })
+            }
+
+            typeMatcher as ClassMatcher
+
+            val pkgMatcher = typeMatcher.pkg
+            val clsMatcher = typeMatcher.classNameMatcher
+            val cp = method.enclosingClass.classpath
+
+            if (pkgMatcher is NameExactMatcher && clsMatcher is NameExactMatcher) {
+                val type = cp.findTypeOrNull("${pkgMatcher.name}$DOT_DELIMITER${clsMatcher.name}")
+                    ?: return mkOr(emptyList())
+                return mkOr(position.map { TypeMatches(it, type) })
+            }
+
+            val alternatives = typeMatcher.extractAlternatives()
+            val disjuncts = mutableListOf<Condition>()
+
+            alternatives.forEach { classMatcher ->
```

Comment:
It is better to use `findTypeOrNull` for resolved alternatives if possible, since `findSubClasses` will load all classes.